### PR TITLE
Paragraph Indentation

### PR DESCRIPTION
**Pull request by metal450**
Implements https://github.com/giuspen/cherrytree/issues/849

I added 2 new controls, "indent" & "unindent," with keyboard shortcuts Ctrl+Shift+> and Ctrl+Shift+<. Pressing them repeatedly increases & decreases the indent of the current paragraph. It seems to be fully working on my end, with two caveats: I didn't implement importing/exporting (i.e. html), and of course, there's still the issue @txe mentioned about "forwards-compatibility" with the old cherry tree. It does properly save/restore in the c++ version.

Hopefully you'll consider merging - I can't wait to get started using ct, & this was the only blocking feature for me (given my current library of content that I plan to migrate in) :)

### CODE REVIEW

**Comment by txe** [comment]
OMG) We will discuss

File: future/src/ct/ct_main_win.cc [hunk 1]
```diff
@@ -224,12 +224,18 @@ Glib::RefPtr<Gsv::Buffer> CtMainWin::get_new_text_buffer(const Glib::ustring& te
...
-const std::string CtMainWin::get_text_tag_name_exist_or_create(const std::string& propertyName, const std::string& propertyValue)
+const std::string CtMainWin::get_text_tag_name_exist_or_create(const std::string& propertyName, const std::string& propertyValue, std::optional<Glib::RefPtr<Gtk::TextBuffer>> text_buffer)
 {
     const std::string tagName{propertyName + "_" + propertyValue};
     Glib::RefPtr<Gtk::TextTag> rTextTag = _rGtkTextTagTable->lookup(tagName);
     if (not rTextTag)
     {
+        if (CtConst::TAG_INDENT == propertyName && text_buffer )
+        {
+            gtk_text_buffer_create_tag((*text_buffer)->gobj(), &(tagName[0]), "left_margin", std::stoi( propertyValue ), NULL);
```
Comment (ForeverRainbow) [inline future/src/ct/ct_main_win.cc:235]: Is this needed? Also there is [Gtk::TextBuffer::create_tag](https://developer.gnome.org/gtkmm/stable/classGtk_1_1TextBuffer.html#a048bdd55d3f244e584074676ab482a2b) :)

Also `NULL` should be replaced by `nullptr` in C++

**Comment by txe** [comment]
I'll commit the changes on Mon or Tue, I postpone a bit to have time to check some stuff.
If you want, you can add `Gtk::TextBuffer::create_tag` or leave it as is.

**Comment by giuspen** [comment]
I will try and packages 0.99.6 in the evening and it is a good idea to leave this out for today so there will be more time to test before 0.99.7

**Comment by metal450** [comment]
> I'll commit the changes on Mon or Tue, I postpone a bit to have time to check some stuff.
> If you want, you can add `Gtk::TextBuffer::create_tag` or leave it as is.

Awesome, thanks!  re create_tag I'll leave that up to you guys if it's your preferred approach - was just doing my best to follow the example of the existing styles & come up with something that behaved as expected, which this did :)

**Comment by txe** [comment]
Sorry that it takes so long.
Code works OK and pygtk doesn't have any issue with a new tag.

I changed a few things:
- instead of keeping the indent in pixels (like 50, 100 and etc), it's now amount of indents (1, 2, 3). It will allows to customize the actual indent later without changing user data. Indent is 50 now, but can be calculated from font size or taken from user input.
- I removed "Line Wrapping Indention" (usually -14) for indented paragraphs. I think this way it looks better.

**Comment by txe** [comment]
Also instead of making zero indent, the code just removes indent tag

**Comment by metal450** [comment]
Agreed - thanks!!

Do you think it's safe to consider this as the implementation that will be included, & to thus start pulling over my data?  I can start in a copy built from source, until 0.99.7.  Or should I wait until 0.99.7 is official first? :)

**Comment by txe** [comment]
Thanks to you, otherwise it would take even longer.
 
I think, it's ok to use. The code is simple, nothing much to change there.
But I didn't add work with clipboard copy/paste, so you will have to manually adjust indent for now.

@giuspen,  do you think it is more or less done, or there is something to add?

**Comment by metal450** [comment]
Yup, I know - much of my formatting needs to be done by hand anyway, so that's OK.  It's a one time operation.  As long as there's some way to have the formatting, which now there is :)

The only thing I think would be nice would be to have this indentation also _exported_ to HTML, but less essential until this has actually been around long enough for people to have started using it ;)

**Comment by metal450** [comment]
Fyi, just a small note: my rationale for storing pixels was for the possibility in the future of having "ruler-style" indentation.  i.e. like in MS Word, where you can drag the margin to any exact offset you want, rather than just to predefined "multiples."  You could, of course, still configure the "hotkey" interval to be something different, but the user could use a ruler to drag the indentation/margin to any precise level, and changing the distance indented by a keypress wouldn't alter existing indents.

For me this is not necessary at all, and is probably overkill, so storing "indentation level" rather than "indentation pixels" as you propose is probably more than enough.  Possibly *slightly* less flexible, though :)

**Comment by metal450** [comment]
(I'm still starting to migrate my data now, so hopefully you don't prefer to change it back - but I just thought I'd share what my thoughts had been, for completeness :) )

**Comment by txe** [comment]
Thanks, good to know. But I like it the current way, I don't want to convert the note-taking program into the word-processing program, though the difference may be not obvious.

But it reminds me where indent also should be used. In bullet/number lists, if the line is too long and wrapped, the wrapped line starts from the left edge of a texview. It would be better to align this line with bullet/number symbol.

**Comment by metal450** [comment]
Absolutely, 100%.  Also relates to: https://github.com/giuspen/cherrytree/issues/935 :)

**Comment by giuspen** [comment]
@txe I haven't looked in detail yet since overworking for a work deadline in my company this Friday but I believe the next step will be to try and support this new indentation in the exports to pdf and html.
certainly this can be used to improve the lists but lower priority I reckon.
I'm not completely satisfied about the svg icons I may replace them but that is a minor.

**Pull request merged**